portfolio/src/main/java/com/google/sps/servlets/ChooseUsernameServlet.java

```diff
@@ -34,21 +39,51 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
     if (!userService.isUserLoggedIn()) {
-      response.sendRedirect("/index.html");
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED,
+          "You cannot set a username if you are not logged in.");
       return;
     }
 
     String username = request.getParameter("username");
     String id = userService.getCurrentUser().getUserId();
 
+    // Build transaction with cross-group functionality
+    // so it can put both a Username and UserInfo entity
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    // Make a UserInfo Entity with retrieved ID and username
-    Entity entity = new Entity("UserInfo", id);
-    entity.setProperty("id", id);
-    entity.setProperty("username", username);
-    // The put() function automatically inserts new data or updates existing data based on ID
-    datastore.put(entity);
+    TransactionOptions options = TransactionOptions.Builder.withXG(true);
+    Transaction transaction = datastore.beginTransaction(options);
+    try {
+      Entity usr = new Entity("Username", username);
+      Entity check = null;
+      try {
+        Key usrkey = KeyFactory.createKey("Username", username);
+        check = datastore.get(usrkey);
+      } catch (EntityNotFoundException e) {
+        // This is what we want (the entity was not found, free to use the username)
+      }
 
+      if (check != null) {
+        // An entity was found with the requested username
+        response.setStatus(409, "Username already exists.");
+        response.getWriter().close();
+        transaction.commit();
```

Comment:
I'm kind of confused by this. It looks like you're trying to commit a failure case?

I'd think the way the flow looks is more like, attempting to commit causes a failure and needs to roll back or something like that.

```
txn = datastore.newTransaction();
try {
    // Check if user already exists.
    entity username = txn.get(username);
    if (username != null) {
        // user already exists; return error code.
    }

    // set our username, whatever that looks like.
    txn.put(user, username);

    // commit our transaction.    
    txn.commit();
} finally {
    // The finally block always executes (whether you're returning, or throwing an exception).
    if (txn.isActive()) {
        // If at this point there's an active transaction, roll it back.
        txn.rollback();
    }
}
```

Have a look at this doc, it gives a closer look at the semantics:

https://cloud.google.com/datastore/docs/concepts/transactions#datastore-datastore-transactional-update-java

Reply:
Great, I think I understand this better now! I have two put() calls in the transaction because I am having one keyed by usernames themselves to check for duplicates and one keyed by user IDs that are also connected with emails. I see now that was not a great idea but it seemed to be the solution I could find where I wouldn't have to change my whole comment structure.